vllm/v1/worker/mps_model_runner.py

```diff
@@ -0,0 +1,146 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""MPS Model Runner for Apple Silicon."""
+
+from contextlib import contextmanager
+from typing import Any
+
+import torch
+import torch.nn as nn
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.model_executor.model_loader import get_model
+from vllm.v1.kv_cache_interface import KVCacheConfig
+from vllm.v1.worker.gpu_model_runner import GPUModelRunner
+
+logger = init_logger(__name__)
+
+
+class MPSModelRunner(GPUModelRunner):
+    """Model runner for MPS (Metal Performance Shaders) on Apple Silicon.
+
+    This runner keeps the model and input tensors on MPS for GPU acceleration,
+    but allocates the KV cache on CPU for numerical stability. The attention
+    backend handles the device movement between MPS and CPU.
+    """
+
+    def __init__(self, vllm_config: VllmConfig, device: torch.device):
+        assert device == torch.device("mps"), f"Expected MPS device, got {device}"
+
+        # Initialize with MPS device - inputs and model will be on MPS
+        with _torch_cuda_wrapper():
+            super().__init__(vllm_config, device)
+
+        assert self.speculative_config is None, "Speculative decoding is not supported on MPS."
+
+        self.use_cuda_graph = False
+        self.cascade_attn_enabled = False
+
+    def load_model(self, eep_scale_up: bool = False) -> None:
+        """Load the model onto MPS device."""
+        logger.info("Starting to load model %s...", self.model_config.model)
+
+        # Load model - it will be loaded on MPS due to device config
+        self.model = get_model(vllm_config=self.vllm_config)
+
+        # Ensure model is on MPS
+        if not all(p.device.type == "mps" for p in self.model.parameters() if p.numel() > 0):
+            logger.info("Moving model to MPS device...")
+            self.model = self.model.to("mps")
```

Comment:
Is this left over from debugging or does the context manager that's specifies device not always work for MPS?